pkg/querier/queryrange/splitters.go

```diff
@@ -0,0 +1,292 @@
+// Large range vector durations will not be split into smaller intervals because it can cause the queries to be slow by over-processing data.
+func (s *metricQuerySplitter) reduceSplitIntervalForRangeVector(r *LokiRequest, interval time.Duration) (time.Duration, error) {
+	maxRange, _, err := maxRangeVectorAndOffsetDuration(r.Plan.AST)
+	if err != nil {
+		return 0, err
+	}
+	if maxRange > interval {
+		return maxRange, nil
+	}
+	return interval, nil
+}
+
+// Round up to the step before the next interval boundary.
+func (s *metricQuerySplitter) nextIntervalBoundary(t time.Time, step int64, interval time.Duration) time.Time {
+	stepNs := step * 1e6
+	nsPerInterval := interval.Nanoseconds()
+	startOfNextInterval := ((t.UnixNano() / nsPerInterval) + 1) * nsPerInterval
+	// ensure that target is a multiple of steps away from the start time
+	target := startOfNextInterval - ((startOfNextInterval - t.UnixNano()) % stepNs)
+	if target == startOfNextInterval {
+		target -= stepNs
+	}
+	return time.Unix(0, target)
+}
+
+func (s *metricQuerySplitter) split(execTime time.Time, tenantIDs []string, r queryrangebase.Request, interval time.Duration) ([]queryrangebase.Request, error) {
+	var reqs []queryrangebase.Request
+
+	lokiReq := r.(*LokiRequest)
+
+	interval, err := s.reduceSplitIntervalForRangeVector(lokiReq, interval)
+	if err != nil {
+		return nil, err
+	}
+
+	start, end := s.alignStartEnd(r.GetStep(), lokiReq.StartTs, lokiReq.EndTs)
+
+	lokiReq = lokiReq.WithStartEnd(start, end).(*LokiRequest)
+
+	factory := func(start, end time.Time) {
+		reqs = append(reqs, &LokiRequest{
+			Query:     lokiReq.Query,
+			Limit:     lokiReq.Limit,
+			Step:      lokiReq.Step,
+			Interval:  lokiReq.Interval,
+			Direction: lokiReq.Direction,
+			Path:      lokiReq.Path,
+			StartTs:   start,
+			EndTs:     end,
+			Plan:      lokiReq.Plan,
+		})
+	}
+
+	// step is >= configured split interval, let us just split the query interval by step
+	// TODO this is likely buggy when step >= query range, how should we handle this?
+	if lokiReq.Step >= interval.Milliseconds() {
+		util.ForInterval(time.Duration(lokiReq.Step*1e6), lokiReq.StartTs, lokiReq.EndTs, false, factory)
+
+		return reqs, nil
+	}
+
+	var (
+		ingesterSplits      []queryrangebase.Request
+		needsIngesterSplits bool
+	)
+
+	origStart := start
+	origEnd := end
+
+	start, end, needsIngesterSplits = ingesterQueryBounds(execTime, s.iqo, lokiReq)
+	start, end = s.alignStartEnd(r.GetStep(), start, end)
+
+	if ingesterQueryInterval := validation.MaxDurationPerTenant(tenantIDs, s.limits.IngesterQuerySplitDuration); ingesterQueryInterval != 0 && needsIngesterSplits {
+		// perform splitting using special interval (`split_ingester_queries_by_interval`)
+		s.buildMetricSplits(lokiReq.GetStep(), ingesterQueryInterval, start, end, factory)
+
+		// rebound after ingester queries have been split out
+		end = start
```

Comment:
we should also substract step from end to leave a step gap on the ingester split boundary similar to what we do with each split.


https://github.com/grafana/loki/blob/b7dfa09cbf74112d4d9a0e572ade02e36207ec36/pkg/querier/queryrange/splitters.go#L155

Reply:
Implemented in https://github.com/grafana/loki/pull/11535/commits/75ee0159621c0eb5d96890fcf7076192a75d2389, good catch!